### PR TITLE
store/postgres: make sure only one graph-node runs database migrations

### PR DESCRIPTION
**Pull request by lutter**
We had the problem that if multiple graph-nodes started up simultaneously,
all of them would try to run database migrations. Only one of them would
succeed, and the others would have their migrations rolled back. That
causes issues with slow or expensive migrations because it forces the
database to do a lot of unnecessary work.

With this change, only one graph-node will migrate the database, all the
others will wait for it to finish by blocking on a new
__graph_node_global_lock table.

### CODE REVIEW

**Comment by lutter** [comment]
I tested this by running `cargo test` while creating/locking `__graph_node_global_lock` in a `psql` session and making sure the tests succeeded once I committed the txn in the `psql` session.

File: store/postgres/src/store.rs [hunk 1]
```diff
@@ -27,23 +27,57 @@ use crate::store_events::StoreEventListener;
...
+    // likely caused by us losing a race to create the table against another
+    // graph-node. If this truly is an error, we will trip over it when
+    // we try to lock the table and report it to the user
+    if let Err(e) = blocking_conn.batch_execute(
+        "create table if not exists \
+         __graph_node_global_lock(id int)",
+    ) {
+        debug!(
+            logger,
+            "creating lock table failed, this is most likely harmless: {:?}", e
```
Comment (Jannis) [inline store/postgres/src/store.rs]: Can we use `; "error" => format!("{:?}", e)` here?

Comment (Jannis) [inline store/postgres/src/store.rs]: Capital `Creating`

File: store/postgres/src/store.rs [hunk 2]
```diff
@@ -27,23 +27,57 @@ use crate::store_events::StoreEventListener;
...
+
+    // blocking_conn holds the lock on the migrations table for the duration
+    // of the migration on conn. Since all nodes execute this code, only one
+    // of them can run this code at the same time. We need to use two
+    // connections for this because diesel will run each migration in its
+    // own txn, which makes it impossible to hold a lock across all of them
+    // on that connection
+    info!(
+        logger,
+        "waiting for other graph-node instances to finish migrating"
```
Comment (Jannis) [inline store/postgres/src/store.rs]: Capital `Waiting`

File: store/postgres/src/store.rs [hunk 3]
```diff
@@ -27,23 +27,57 @@ use crate::store_events::StoreEventListener;
...
+    );
+    let result = blocking_conn.transaction(|| {
+        diesel::sql_query("lock table __graph_node_global_lock in exclusive mode")
+            .execute(blocking_conn)?;
+        info!(logger, "running migrations");
```
Comment (Jannis) [inline store/postgres/src/store.rs]: Capital `Running`

File: store/postgres/src/store.rs [hunk 4]
```diff
@@ -27,23 +27,57 @@ use crate::store_events::StoreEventListener;
...
+        embedded_migrations::run_with_output(conn, &mut output)
+    });
+    info!(logger, "migrations finished");
```
Comment (Jannis) [inline store/postgres/src/store.rs]: Capital `Migrations`

**Comment by lutter** [comment]
Rebased to latest master and made the requested changes

**Pull request merged**